### PR TITLE
RSDK-6588 - Make background camera loop

### PR DESCRIPTION
**Pull request by kharijarrett**
This came together finally. Here are some things you won't see in this PR:

1. Memory management (aka dumping tracks that need dumping)
2. Speed management (aka using motion blobs > detections some frames)
3. Tracking improvement (aka actually adding the feedback for Kalman filter)

### CODE REVIEW

File: object_tracker/object_tracker.go [hunk 1]
```diff
@@ -15,10 +18,12 @@ import (
...
+	viamutils "go.viam.com/utils"
 	"image"
 )
 
-// ModelName is the name of the model
+// ModelName is the name of the model, MinConfidence is the minimum confidence of a detection
+// worth tracking, and DefaultMaxFrequency (in Hz) is the frequency of the bkgnd loop
 const (
 	ModelName           = "object-tracker"
 	MinConfidence       = 0.2
```
Comment (bhaney) [inline object_tracker/object_tracker.go]: I honestly think that this should be a parameter in the config -- some models have really wild confidences (like I remember seeing a person model that you HAD to set at .95 because .5 confidence would draw boxes around empty patches of walls)

File: object_tracker/object_tracker.go [hunk 2]
```diff
@@ -48,57 +71,108 @@ func newTracker(ctx context.Context, deps resource.Dependencies, conf resource.C
...
 		return nil, err
 	}
 	for i := 0; i < 2; i++ {
-		img, _, err := stream.Next(ctx)
+		img, _, err := stream.Next(t.cancelContext)
 		if err != nil {
 			return nil, err
 		}
-		detections, err := t.detector.Detections(ctx, img, nil)
+		detections, err := t.detector.Detections(t.cancelContext, img, nil)
```
Comment (bhaney) [inline object_tracker/object_tracker.go]: So, I think these two detections should just use the constructor's context ... my reasoning is that if for some reason the initialization of the model has to be stopped, it's going to use the `ctx` from the constructor. But this is the only place where `ctx` should be used rather than cancelCtx, i believe.

File: object_tracker/object_tracker.go [hunk 3]
```diff
@@ -48,57 +71,108 @@ func newTracker(ctx context.Context, deps resource.Dependencies, conf resource.C
...
+			start := time.Now()
+			// Load up the old detections
+			namedOld := t.oldDetections.Load()[1]
+
+			// Take fresh detections from fresh image
+			img, _, err := stream.Next(streamCtx)
+			if err != nil {
+				return
+			}
+			detections, err := t.detector.Detections(streamCtx, img, nil)
```
Comment (bhaney) [inline object_tracker/object_tracker.go]: Is there a reason you are using a new background context rather than the `cancelableCtx` here?

File: object_tracker/object_tracker.go [hunk 4]
```diff
@@ -48,57 +71,108 @@ func newTracker(ctx context.Context, deps resource.Dependencies, conf resource.C
...
 	if err != nil {
 		return nil, err
 	}
 	matches := HA.Execute()
-
 	// Rename from temporal matches. New det copies old det's label
 	renamedNew := t.RenameFromMatches(matches, renamedOld, filteredNew)
-	t.oldDetections[0] = renamedOld
-	t.oldDetections[1] = renamedNew
+	t.oldDetections.Store(&[2][]objdet.Detection{renamedOld, renamedNew})
```
Comment (bhaney) [inline object_tracker/object_tracker.go:117]: You need to be careful here with shared memory -- created a `mutex` object in the struct, and every time you are about to read or write from these detections, mutexLock/Unlock around it.

File: object_tracker/object_tracker.go [hunk 5]
```diff
@@ -48,57 +71,108 @@ func newTracker(ctx context.Context, deps resource.Dependencies, conf resource.C
...
+
+			// Build and solve cost matrix via Munkres' method
+			matchMtx := t.BuildMatchingMatrix(namedOld, filteredNew)
+			HA, _ := hg.NewHungarianAlgorithm(matchMtx)
+			matches := HA.Execute()
+			// Rename from temporal matches. New det copies old det's label
+			renamedNew := t.RenameFromMatches(matches, namedOld, filteredNew)
+
+			// Store the matched detections
+			t.oldDetections.Store(&[2][]objdet.Detection{namedOld, renamedNew})
```
Comment (bhaney) [inline object_tracker/object_tracker.go:164]: Need a mutex Lock/Unlock around this

File: object_tracker/object_tracker.go [hunk 6]
```diff
@@ -175,63 +234,24 @@ func (t *myTracker) DetectionsFromCamera(
...
 		return nil, errors.Errorf("Camera name given to method, %v is not the same as configured camera %v", cameraName, t.camName)
 	}
-	stream, err := t.cam.Stream(ctx, nil)
-	if err != nil {
-		return nil, err
+	select {
+	case <-t.cancelContext.Done():
+		return nil, t.cancelContext.Err()
+	default:
+		return t.oldDetections.Load()[1], nil
```
Comment (bhaney) [inline object_tracker/object_tracker.go:257]: Use a mutex Read lock/unlock around this -- i.e. load into a new detection object, and return that new detection object

File: object_tracker/object_tracker.go [hunk 7]
```diff
@@ -175,63 +234,24 @@ func (t *myTracker) DetectionsFromCamera(
...
-	// Start by grabbing the old detections. They're filtered and renamed already.
-	namedOld := t.oldDetections[1]
-
-	// Take fresh detections
-	detections, err := t.detector.Detections(ctx, img, nil)
-	if err != nil {
-		return nil, err
+	select {
+	case <-t.cancelContext.Done():
+		return nil, t.cancelContext.Err()
```
Comment (bhaney) [inline object_tracker/object_tracker.go:264]: check for the context.Done of the ctx from the arguments, too
```
		case <-t.cancelContext.Done():
			return nil, t.cancelContext.Err()
		case <-ctx.Done():
			return nil, ctx.Err()
               default:
```

File: object_tracker/object_tracker.go [hunk 8]
```diff
@@ -175,63 +234,24 @@ func (t *myTracker) DetectionsFromCamera(
...
-
 	if cameraName != t.camName {
 		return nil, errors.Errorf("Camera name given to method, %v is not the same as configured camera %v", cameraName, t.camName)
 	}
-	stream, err := t.cam.Stream(ctx, nil)
-	if err != nil {
-		return nil, err
+	select {
+	case <-t.cancelContext.Done():
+		return nil, t.cancelContext.Err()
```
Comment (bhaney) [inline object_tracker/object_tracker.go:253]: check for the context.Done of the ctx from the arguments, too
```
		case <-t.cancelContext.Done():
			return nil, t.cancelContext.Err()
		case <-ctx.Done():
			return nil, ctx.Err()
               default:
```

**Review by bhaney — CHANGES_REQUESTED** [review]
Nice work! Just need to add some mutex stuff, and I have some questions about the contexts used

File: object_tracker/object_tracker.go (hunk 6 above)
Comment (bhaney) [reply object_tracker/object_tracker.go:257]: JK! Not necessary when using atomic, like you are doing

File: object_tracker/object_tracker.go (hunk 5 above)
Comment (bhaney) [reply object_tracker/object_tracker.go:164]: actually you don't need a mutex

File: object_tracker/object_tracker.go (hunk 4 above)
Comment (bhaney) [reply object_tracker/object_tracker.go:117]: no mutex needed

File: object_tracker/object_tracker.go [hunk 9]
```diff
@@ -48,57 +71,108 @@ func newTracker(ctx context.Context, deps resource.Dependencies, conf resource.C
...
+		default:
+			streamCtx := context.Background()
+			start := time.Now()
+			// Load up the old detections
+			namedOld := t.oldDetections.Load()[1]
+
+			// Take fresh detections from fresh image
+			img, _, err := stream.Next(streamCtx)
+			if err != nil {
+				return
```
Comment (bhaney) [inline object_tracker/object_tracker.go:147]: Also, if your err != nil, you really should still at least print what the error is (to the logger)

File: object_tracker/object_tracker.go [hunk 10]
```diff
@@ -48,57 +71,108 @@ func newTracker(ctx context.Context, deps resource.Dependencies, conf resource.C
...
+			}
+			detections, err := t.detector.Detections(streamCtx, img, nil)
+			if err != nil {
+				return
```
Comment (bhaney) [inline object_tracker/object_tracker.go:152]: Same here- print the error to the logger before returning

File: object_tracker/object_tracker.go [hunk 11]
```diff
@@ -258,10 +278,12 @@ func (t *myTracker) GetObjectPointClouds(
 }
 
 func (t *myTracker) Close(ctx context.Context) error {
```
Comment (bhaney) [inline object_tracker/object_tracker.go:295]: I think you need to close your, t.cam.Stream() as well, but it should be done associated with your run() function

Comment (bhaney) [reply object_tracker/object_tracker.go:295]: Oh I see you do it in the ManagedGo function! Nevermind!

File: object_tracker/object_tracker.go (hunk 3 above)
Comment (kharijarrett) [reply object_tracker/object_tracker.go]: No. This is leftover from testing

File: object_tracker/object_tracker.go (hunk 2 above)
Comment (kharijarrett) [reply object_tracker/object_tracker.go]: Okay yeah I updated the detection ctx (but not the stream)

File: object_tracker/object_tracker.go (hunk 1 above)
Comment (kharijarrett) [reply object_tracker/object_tracker.go]: Very fair. Changed it to an optional parameter. Also tested and made sure 0 is different from empty (you'll get default = 0.2)

Comment (kharijarrett) [reply object_tracker/object_tracker.go]: I'll update the readme with the config and default info

File: object_tracker/object_tracker.go [hunk 12]
```diff
@@ -122,6 +122,7 @@ func newTracker(ctx context.Context, deps resource.Dependencies, conf resource.C
 	}, func() {
 		t.cancelFunc()
 		stream.Close(t.cancelContext)
+		ctx.Done()
```
Comment (bhaney) [inline object_tracker/object_tracker.go]: You don't need to do this. This context is for building the module in the first place. You don't need to keep it around. It will become irrelevant once `newTracker` exits

**Review by bhaney — CHANGES_REQUESTED** [review]
Nice! Very close to being done, just this one change

File: object_tracker/object_tracker.go (hunk 12 above)
Comment (bhaney) [reply object_tracker/object_tracker.go]: Also, ctx.Done() is a `channel` type, so I'm not sure if this syntax is correct either (are you trying to receive a value for this channel or wait until it's done? You would to use `<-ctx.Done()` for that)

**Review by bhaney — APPROVED** [review]
LGTM
